aea/mail/base.py

```diff
@@ -392,39 +369,39 @@ def __init__(
                 bool(protocol_id) or bool(protocol_specification_id),
                 "Message is bytes object, protocol_id or protocol_specification_id must be provided!",
             )
+            if protocol_id is not None and protocol_specification_id is not None:
+                protocol_specification_id_ = self._get_protocol_specification_id_by_the_protocol_id(
+                    protocol_id
+                )
+                enforce(
+                    protocol_specification_id == protocol_specification_id_,
+                    "Protocol specification id mismatch.",
+                )
+            elif protocol_id is not None:
+                protocol_specification_id = self._get_protocol_specification_id_by_the_protocol_id(
+                    protocol_id
+                )
         elif isinstance(message, Message):
-            # protocol_id provided as an argument in priority
-            # use Message.protocol_id only if no protocol_id provided
             if message.protocol_id is None:
                 raise ValueError(  # pragma: nocover
                     f"message class {type(message)} has no protocol_id specified!"
                 )
+            if protocol_id is not None:
+                enforce(
+                    protocol_id == message.protocol_id, "Inconsistent args provided."
+                )
             protocol_id = message.protocol_id
+            protocol_specification_id = self._get_protocol_specification_id_by_the_protocol_id(
+                protocol_id
+            )
         else:
             raise ValueError(
                 f"Message type: {type(message)} is not supported!"
             )  # pragma: nocover
 
-        if not protocol_id:
-            protocol_id = self._get_protocol_id_by_protocol_specification_id(
-                cast(PublicId, protocol_specification_id)
-            )
-        if not protocol_specification_id:
-            # if no protocol_spcification_id, try to resolve by protocol_id
-            protocol_specification_id = self._get_protocol_specification_id_by_the_protocol_id(
-                protocol_id
-            )
-
-        enforce(
-            protocol_id is not None, "protocol_id was not specified or auto resolved"
-        )
-        enforce(
-            protocol_specification_id is not None,
-            "protocol_specification_id was not specified or auto resolved",
-        )
-
-        self._protocol_id: PublicId = protocol_id
-        self._protocol_specification_id: PublicId = protocol_specification_id
+        # if protocol_specification_id is None then resolution did not yield anything
```

Comment:
Needs to be fixed and tests which fail need to be fixed too. We want `protocol_specification_id` to not be None at this point, otherwise we raise!